File: baekjoon/dp/1932/triangle.ts
const lines: string[] = require('fs').readFileSync(0).toString().split('\n');
const size: number = Number(lines[0]);
var dp: Array<Array<number>> = new Array(size);
var triangle: Array<Array<number>> = new Array(size);

for (let y = 0; y < size; ++y)
{
    dp[y] = new Array<number>(y + 1).fill(0);
    triangle[y] = lines[y + 1].split(' ').map(Number);
}

dp[0][0] = triangle[0][0];
for (let curY = 0; curY < size - 1; ++curY)
{
    let nextY = curY + 1;

    for (let curX = 0; curX <= curY; ++curX)
    {
        for (let i = 0; i <= 1; ++i)
        {
            let nextX = curX + i;

            dp[nextY][nextX] = Math.max(dp[nextY][nextX],
                dp[curY][curX] + triangle[nextY][nextX]
            );
        }
    }
}

console.log(Math.max(...dp[size - 1]));

File: baekjoon/dp/1932/triangle.py
import sys


def get_next_room(cur_x: int) -> list[int]:
    return [cur_x, cur_x + 1]


if __name__ == '__main__':
    def readline(): return sys.stdin.readline()
    def splitIntoInt(): return map(int, readline().split())

    size = int(readline())
    triangle = [list(splitIntoInt()) for _ in range(size)]
    dp = [[0] * i for i in range(1, size + 1)]

    dp[0][0] = triangle[0][0]
    for cur_y in range(size - 1):
        next_y = cur_y + 1
        for cur_x in range(cur_y + 1):
            for next_x in get_next_room(cur_x):
                dp[next_y][next_x] = max(dp[next_y][next_x],
                                         dp[cur_y][cur_x] + triangle[next_y][next_x])

    print(max(dp[size - 1]))
